support UDFs

pgrivachev opened the pull request:
Added support for persistent user-defined functions (inline is not supported).

fivetran-dmitryboykov commented on server/src/test/ZetaSqlWrapper.spec.ts:
```diff
@@ -172,12 +176,38 @@ describe('ZetaSqlWrapper', () => {
+    const udfsCatalog = rootCatalog.catalog?.find(c => c.name === 'udfs');
+    assert.ok(udfsCatalog);
+    const func = udfsCatalog.customFunction?.find(c => c.namePath && arraysAreEqual(c.namePath, ['udfs', 'func']));
+    assert.ok(func);
```
Do we need to check registered arguments, return type?

pgrivachev replied:
Yes, added asserts.

fivetran-dmitryboykov reviewed (changes requested):
Some notes and 2 questions:
1. If we can extract udf’s from AST analyse result?
2. Do we need to remove udf’s from catalog?

pgrivachev commented on server/src/ZetaSqlWrapper.ts:
```diff
@@ -118,6 +122,65 @@ export class ZetaSqlWrapper {
+  }
+
+  registerUdf(udf: Udf): void {
+    if (udf.nameParts.length <= 1) {
```
Looks like it is not possible to have UDF with only name without dataset. At least I don't know how to handle it)

pgrivachev commented on server/src/bigquery/BigQueryClient.ts:
```diff
@@ -49,4 +64,98 @@ export class BigQueryClient implements DbtDestinationClient {
+
+  static toTypeKind(bigQueryTypeKind?: BigQueryTypeKind): TypeKind {
+    if (!bigQueryTypeKind) {
+      return TypeKind.TYPE_UNKNOWN;
```
Do you want to suggest something? The code is:
```typescript
default:
  return TypeKind.TYPE_UNKNOWN;
```

pgrivachev commented on server/src/ZetaSqlWrapper.ts:
```diff
@@ -118,6 +122,65 @@ export class ZetaSqlWrapper {
+    const udfOwner = this.ensureUdfOwnerCatalogExists(udf);
+
+    if (udfOwner.customFunction?.some(c => c.namePath && arraysAreEqual(c.namePath, udf.nameParts))) {
+      return;
```
Seems like enough, let's try.

pgrivachev commented:
> Some notes and 2 questions:
> 
> 1. If we can extract udf’s from AST analyse result?
> 2. Do we need to remove udf’s from catalog?

1. I will check it, probably yes.
2. We can, but I don't see why we should do it for now.

pgrivachev commented:
We can't get information from AST about UDFs because if we don't register UDFs we don't have AST. Also I found one issue, fixed it and added additional tests.
